ignite/evaluator.py

```diff
@@ -0,0 +1,79 @@
+import time
+
+from torch.autograd import Variable
+
+from ignite.history import History
+from ignite.engine import Engine
+from ignite._utils import _to_hours_mins_secs
+
+from enum import Enum
+
+
+class EvaluationEvents(Enum):
+    EVALUATION_STARTING = "evaluation_starting"
+    EVALUATION_COMPLETED = "evaluation_completed"
+    EVALUATION_ITERATION_STARTED = "evaluation_iteration_started"
+    EVALUATION_ITERATION_COMPLETED = "evaluation_iteration_completed"
+    EXCEPTION_RAISED = "exception_raised"
+
+
+class Evaluator(Engine):
+    def __init__(self, inference_function):
+        super(Evaluator, self).__init__(EvaluationEvents)
+
+        self._inference_function = inference_function
+        self.history = History()
+        self.current_evaluation_iteration = 0
+        self.should_terminate = False
+
+    def run(self, dataset):
+        """ Evaluates the dataset"""
+        if self._inference_function is None:
+            raise ValueError("Evaluator must have an inference function in order to evaluate")
+
+        self.current_evaluation_iteration = 0
+        self._fire_event(EvaluationEvents.EVALUATION_STARTING)
+        start_time = time.time()
+
+        for _, batch in enumerate(dataset, 1):
+            self._fire_event(EvaluationEvents.EVALUATION_ITERATION_STARTED)
+            step_result = self._inference_function(batch)
+            if step_result is not None:
+                self.history.append(step_result)
+
+            self.current_evaluation_iteration += 1
+            self._fire_event(EvaluationEvents.EVALUATION_ITERATION_COMPLETED)
+            if self.should_terminate:
+                break
+
+        time_taken = time.time() - start_time
+        hours, mins, secs = _to_hours_mins_secs(time_taken)
+        self._logger.info("Evaluation Complete. Time taken: %02d:%02d:%02d", hours, mins, secs)
+
+        self._fire_event(EvaluationEvents.EVALUATION_COMPLETED)
+
+
+def create_supervised(model, cuda=False):
+    """
+    Factory function for creating an evaluator for supervised models
+
+    Args:
+        model (torch.nn.Module): the model to train
+        cuda (bool, optional): whether or not to transfer batch to GPU (default: False)
+
+    Returns:
+        Trainer: a trainer instance with supervised inference function
+    """
+    def _prepare_batch(batch):
+        x, y = batch
+        if cuda:
+            x, y = x.cuda(), y.cuda()
+        return Variable(x, volatile=True), Variable(y, volatile=True)
```

Comment:
In case if `x` and `y` are not directly tensors, may be better like this (similar to `collate_fn` from DataLoader):
```python
def apply_variable(batch, **variable_kwargs):
    if torch.is_tensor(batch):
        return Variable(batch, **variable_kwargs)
    elif isinstance(batch, string_classes):
        return batch
    elif isinstance(batch, Mapping):
        return {k: apply_variable(sample, **variable_kwargs) for k, sample in batch.items()}
    elif isinstance(batch, Sequence):
        return [apply_variable(sample, **variable_kwargs) for sample in batch]
    else:
        raise TypeError(("batch must contain tensors, numbers, dicts or lists; found {}"
                         .format(type(batch[0]))))
```